### ISSUE TITLE
Bugs in caf::uri

### ISSUE THREAD

**Issue by alatdneg** [actor-framework/actor-framework#1013]
```
caf::uri path = caf::uri_builder().scheme("file").host("localhost").path("/tmp/test/test.{:04d}.exr").make()
std::cout << to_string(path) << std::endl;
std::cout << to_string(*caf::make_uri(to_string(path))) << std::endl;
```
returns
file://localhost/tmp/test/test.{%3A04d}.exr
the second cout segfaults.

As far as I'm aware "{ }" aren't valid url characters and should be % escaped. I suspect that is the root of the problem.

Also it doesn't seem to support 
"file:///" prefix, which is valid AFAIK.

cheers.

**Comment by Neverlord** [comment]
Thanks for reporting these! `file:///` is definitely valid.

**Comment by Neverlord** [comment]
After looking into it, I found that `{` and `}` are not listed as reserved characters [by the RFC](https://tools.ietf.org/html/rfc3986#page-12). They aren't listed as unreserved either, though. I've fixed it by considering every character not explicitly reserved by the standard as unreserved. Some URI implementations choose to percent-escape `{` and `}` while others don't. I couldn't find a good reason for percent-encoding them, so CAF will just treat them as regular characters.

One note on `file:///`: CAF accepts this input string after my fix. However, printing it back will show `file:/`. This is because CAF does not distinguish between an empty authority component (`file:///`) and no authority component (`file:/`). I couldn't think of a scenario where this would ever matter, though.
